comp/netflow/goflowlib/netflowstate/netflow.go

```diff
@@ -0,0 +1,383 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package netflowstate provides a Netflow state manager
+// on top of goflow default producer, to allow additional fields collection.
+package netflowstate
+
+import (
+	"bytes"
+	"context"
+	"github.com/DataDog/datadog-agent/comp/netflow/common"
+	"github.com/DataDog/datadog-agent/comp/netflow/config"
+	"github.com/DataDog/datadog-agent/comp/netflow/goflowlib/additionalfields"
+	"github.com/netsampler/goflow2/utils"
+	"sync"
+	"time"
+
+	"github.com/netsampler/goflow2/decoders/netflow"
+	"github.com/netsampler/goflow2/decoders/netflow/templates"
+	"github.com/netsampler/goflow2/format"
+	flowmessage "github.com/netsampler/goflow2/pb"
+	"github.com/netsampler/goflow2/producer"
+	"github.com/netsampler/goflow2/transport"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// StateNetFlow holds a NetflowV9/IPFIX producer
+type StateNetFlow struct {
+	stopper
+
+	Format    format.FormatInterface
+	Transport transport.TransportInterface
+	Logger    utils.Logger
+
+	samplinglock *sync.RWMutex
+	sampling     map[string]producer.SamplingRateSystem
+
+	Config       *producer.ProducerConfig
+	configMapped *producer.ProducerConfigMapped
+
+	TemplateSystem templates.TemplateInterface
+
+	ctx context.Context
+
+	mappedFieldsConfig map[uint16]config.Mapping
+}
+
+// NewStateNetFlow initializes a new Netflow/IPFIX producer, with the goflow default producer and the additional fields producer
+func NewStateNetFlow(mappingConfs []config.Mapping) *StateNetFlow {
+	return &StateNetFlow{
+		ctx:                context.Background(),
+		samplinglock:       &sync.RWMutex{},
+		sampling:           make(map[string]producer.SamplingRateSystem),
+		mappedFieldsConfig: mapFieldsConfig(mappingConfs),
+	}
+}
+
+// DecodeFlow decodes a flow into common.FlowMessageWithAdditionalFields
+func (s *StateNetFlow) DecodeFlow(msg interface{}) error {
```

Comment:
I don't see `StateNetFlow` in goflow2 v2. Do you know how this customized `StateNetFlow` will evolve when migrating to goflow2 v2?

Reply:
Yes! Goflow2 v2 uses the concept of [pipes](https://github.com/netsampler/goflow2/blob/a0660824f039ab53551d63201985f8293e2c2d4b/utils/pipe.go#L23-L29). The [pipe config](https://github.com/netsampler/goflow2/blob/a0660824f039ab53551d63201985f8293e2c2d4b/utils/pipe.go#L31C1-L37C2) allows passing producer functions, which will allow us to collect custom fields.